crates/wasmtime/src/func.rs

```diff
@@ -360,20 +250,18 @@ macro_rules! generate_wrap_async_func {
         #[cfg(feature = "async")]
         #[cfg_attr(nightlydoc, doc(cfg(feature = "async")))]
         pub fn [<wrap $num _async>]<T, $($args,)* R>(
-            store: &Store,
-            state: T,
-            func: impl for<'a> Fn(Caller<'a>, &'a T, $($args),*) -> Box<dyn Future<Output = R> + 'a> + 'static,
+            store: impl AsContextMut<Data = T>,
+            func: impl for<'a> Fn(Caller<'a, T>, $($args),*) -> Box<dyn Future<Output = R> + Send + 'a> + Send + Sync + 'static,
         ) -> Func
         where
-            T: 'static,
             $($args: WasmTy,)*
             R: WasmRet,
         {
-            assert!(store.async_support(), concat!("cannot use `wrap", $num, "_async` without enabling async support on the config"));
-            Func::wrap(store, move |caller: Caller<'_>, $($args: $args),*| {
-                let store = caller.store().clone();
-                let mut future = Pin::from(func(caller, &state, $($args),*));
-                match store.block_on(future.as_mut()) {
+            assert!(store.as_context().async_support(), concat!("cannot use `wrap", $num, "_async` without enabling async support on the config"));
+            Func::wrap(store, move |mut caller: Caller<'_, T>, $($args: $args),*| {
+                let async_cx = caller.store.as_context_mut().opaque().async_cx();
+                let mut future = Pin::from(func(caller, $($args),*));
+                match unsafe { async_cx.block_on(future.as_mut()) } {
```

Comment:
Note that the async bits here are a little different from before, but I think in a way that makes things a bit clearer. The `async_cx` is a "separately owned" structure which is connected (via unsafety) but not lifetimes to the original `caller`. This allows us to pass the entire `caller` to the future and let it close over it. The Wasmtime implementation, though, has to be careful that the fields referenced by `async_cx` may be aliased, hence their `UnsafeCell` wrappers.